scripts/rebase.sh

```diff
@@ -155,45 +88,119 @@ download_release() {
-        REPO=$(echo "${line}" | cut -d ' ' -f 2)
-        if [[ "${EMBEDDED_COMPONENTS}" == *"${COMPONENT}"* ]]; then
-            extract_section "${REPO##*/}/go.mod" require > require
-            extract_section "${REPO##*/}/go.mod" replace > replace
-            update_versions latest_require require > t; mv t latest_require
-            update_versions latest_replace replace > t; mv t latest_replace
-        fi
-    done < source-commits
+#
+get_modulepath_version_for_release() {
+    local component=$1
 
-    cat << EOF > "${REPOROOT}/go.mod"
-module github.com/openshift/microshift
+    repo=$( cd "${STAGING_DIR}/${component}" && git config --get remote.origin.url )
+    modulepath="${repo#https://}"
+    echo "${modulepath}@$(get_pseudoversion ${modulepath} ${component})"
+}
 
-go 1.16
+# Returns the line (including trailing comment) in the #{gomod_file} containing the ReplaceDirective for ${module_path}
+get_replace_directive() {
+    local gomod_file=$1
+    local module_path=$2
 
-replace (
-$(cat latest_replace)
-)
+    # TODO: Handle special case of keyword "replace" being included in the line
+    go mod edit -print "${gomod_file}" | grep "^[[:space:]]${module_path}[[:space:]][[:alnum:][:space:].-]*=>"
+}
 
-require (
-$(cat latest_require)
-)
-EOF
+lookup_modulepath_version_from_component() {
+    local modulepath=$1
+    local component=$2
+
+    # special-case etcd
+    if [[ "${modulepath}" =~ ^go.etcd.io/etcd/ ]]; then
+        modulepath=$(echo "${modulepath}" | sed 's|^go.etcd.io/etcd|github.com/openshift/etcd|')
+        pseudoversion=$(get_pseudoversion ${modulepath} etcd)
+        echo "${modulepath}@${pseudoversion}"
+        return
+    fi
+
+    replace_directive=$(get_replace_directive "${STAGING_DIR}/${component}/go.mod" "${modulepath}")
+    replace_directive=$(strip_comment "${replace_directive}")
+    replacement=$(echo "${replace_directive}" | sed -E "s|.*=>[[:space:]]*(.*)[[:space:]]*|\1|")
+    if [[ "${replacement}" =~ ^./staging ]]; then
+        replacement=$(echo "${replacement}" | sed 's|^./staging/|github.com/openshift/kubernetes/staging/|')
+        replacement="${replacement} $(get_pseudoversion ${modulepath} kubernetes)"
+    fi
+    echo "${replacement}" | sed 's| |@|'
+}
+
+# Returns ${line} without comment
+strip_comment() {
+    local line=$1
+
+    echo "${line%%//*}"
+}
 
-    go mod tidy
-    go mod vendor
+# Returns the comment in ${line} if one exists or an empty string if not
+get_comment() {
+    local line=$1
+
+    comment=${line##*//}
+    if [ "${comment}" != "${line}" ]; then
+        echo ${comment}
+    else
+        echo ""
+    fi
+}
+
+valid_component_or_exit() {
+    local component=$1
+    if [[ ! " etcd kubernetes openshift-apiserver openshift-controller-manager " =~ " ${component} " ]]; then
+        echo "error: release reference must be one of [etcd kubernetes openshift-apiserver openshift-controller-manager], have ${component}"
+        exit 1
+    fi
+}
+
+update_go_mod() {
```

Comment:
When I run `./scripts/rebase.sh go.mod` I get this all the way down: 
```
$ ./scripts/rebase.sh go.mod
# Updating go.mod
skipping modulepath bazil.org/fuse: no or unknown command []
skipping modulepath bitbucket.org/bertimus9/systemstat: no or unknown command []
skipping modulepath cloud.google.com/go: no or unknown command []
skipping modulepath cloud.google.com/go/bigquery: no or unknown command []
---
```
and this returns empty from command line, also:
`replaced_modulepaths=$(go mod edit -json | jq -r '.Replace // []' | jq -r '.[].Old.Path' | xargs)` 

The main branch rebase.sh works as expected.

Reply:
> When I run `./scripts/rebase.sh go.mod` I get this all the way down:
This is expected as you're likely using a go.mod file that isn't annotated yet (what PR would address).
Basically, if we want more predictability when resolving conflicts between replace directives of different embedded components, we need a way to tell the rebase script how to. The annotations do that explicitly and that hopefully is more stable than "choose latest", which may change during a release.

> and this returns empty from command line, also: `replaced_modulepaths=$(go mod edit -json | jq -r '.Replace // []' | jq -r '.[].Old.Path' | xargs)`
That's weird. I'm not seeing that on the command line. And in the script at least it should work for you, too, otherwise you'd not be seeing all these "skipping..." messages.

Reply:
and, after a laptop restart all is well.